src/redis-cli.c

```diff
@@ -706,48 +706,67 @@ static void completionCallback(const char *buf, linenoiseCompletions *lc) {
 static char *hintsCallback(const char *buf, int *color, int *bold) {
     if (!pref.hints) return NULL;
 
-    int i, argc, buflen = strlen(buf);
-    sds *argv = sdssplitargs(buf,&argc);
+    int i, rawargc, argc, buflen = strlen(buf), matchlen = 0;
+    sds *rawargv, *argv = sdssplitargs(buf,&argc);
     int endspace = buflen && isspace(buf[buflen-1]);
+    helpEntry *entry = NULL;
 
     /* Check if the argument list is empty and return ASAP. */
     if (argc == 0) {
         sdsfreesplitres(argv,argc);
         return NULL;
     }
 
+    /* Search longest matching prefix command */
     for (i = 0; i < helpEntriesLen; i++) {
         if (!(helpEntries[i].type & CLI_HELP_COMMAND)) continue;
 
-        if (strcasecmp(argv[0],helpEntries[i].full) == 0 ||
-            strcasecmp(buf,helpEntries[i].full) == 0)
-        {
-            *color = 90;
-            *bold = 0;
-            sds hint = sdsnew(helpEntries[i].org->params);
-
-            /* Remove arguments from the returned hint to show only the
-             * ones the user did not yet typed. */
-            int toremove = argc-1;
-            while(toremove > 0 && sdslen(hint)) {
-                if (hint[0] == '[') break;
-                if (hint[0] == ' ') toremove--;
-                sdsrange(hint,1,-1);
+        rawargv = sdssplitargs(helpEntries[i].full,&rawargc);
+        if (rawargc <= argc) {
+            int j;
+            for (j = 0; j < rawargc; j++) {
+                if (strcasecmp(rawargv[j],argv[j])) {
+                    break;
+                }
             }
-
-            /* Add an initial space if needed. */
-            if (!endspace) {
-                sds newhint = sdsnewlen(" ",1);
-                newhint = sdscatsds(newhint,hint);
-                sdsfree(hint);
-                hint = newhint;
+            if (j == rawargc && rawargc > matchlen) {
+                matchlen = rawargc;
+                entry = &helpEntries[i];
             }
-
-            sdsfreesplitres(argv,argc);
-            return hint;
         }
+        sdsfreesplitres(rawargv,rawargc);
     }
     sdsfreesplitres(argv,argc);
+
+    if (entry) {
+        /* Command with no params should return ASAP */
+        if (strlen(entry->org->params) == 1 && entry->org->params[0] == '-') {
+            return NULL;
+        }
```

Comment:
i wonder why at all we need that `-` in help.h.
when doing `help client` it shows too, but it seems completely useless to me.
i tried looking at the history of the generator script, and it's there since the beginning.
maybe we need to remove it and re-generate help.h?

Reply:
so we can hide it from the `HELP` thing too.
i.e.
```
127.0.0.1:6379> help client list

  CLIENT LIST -
  summary: Get the list of client connections
  since: 2.4.0
  group: server
```
or instead, just remove it from `utils/generate-command-help.rb` and re-generate help.h.
i think i prefer to delete it if we don't see any reason it would be here.

Reply:
If hide it from the `HELP`, I think we should delete it. Because it will be useless.

Reply:
@yoav-steinberg I think I should delete these lines.

Reply:
i made a separate PR to trim that off. https://github.com/redis/redis/pull/8939
so we can now just remove the 3 new lines from this PR.